### PR TITLE
Simple validation of the package name in the uploaded pubspec.

### LINKED ISSUE #50: Validate package name in the backend
Currently the pub client validates package names (see https://github.com/dart-lang/pub/blob/master/lib/src/validator/name.dart#L51). However the backend does not, so if packages are uploaded by some other means than the pub client (or using a modified pub client) illegal package names will be accepted.

### PR DESCRIPTION
**Pull request by isoos**
Fixes #50.

### CODE REVIEW

File: app/lib/utils.dart [hunk 1]
```diff
@@ -51,3 +51,56 @@ String canonicalizeVersion(String version) {
   }
   return canonicalVersion.toString();
 }
+
+RegExp _identifierExpr = new RegExp(r'^[a-zA-Z0-9_]*$');
+RegExp _startsWithLetterOrUnderscore = new RegExp(r'^[a-zA-Z_]');
```
Comment (mkustermann) [inline app/lib/utils.dart]: Consider always using `final` if you don't modify variables (also local variables)

File: app/lib/utils.dart [hunk 2]
```diff
@@ -51,3 +51,56 @@ String canonicalizeVersion(String version) {
...
+const _reservedWords = const [
```
Comment (mkustermann) [inline app/lib/utils.dart]: Consider also typing it `const <String> [ ... ]`. Now that we are moving towards the strong mode type system, it's probably better.

File: app/test/backend_test.dart [hunk 3]
```diff
@@ -731,6 +731,37 @@ main() {
...
 
+        scopedTest('bad package names are rejected', () async {
+          var tarballStorage = new TarballStorageMock();
+          var transactionMock = new TransactionMock();
+          var db = new DatastoreDBMock(transactionMock: transactionMock);
+          var repo = new GCloudPackageRepository(db, tarballStorage);
+          registerLoggedInUser('hans@juergen.com');
+
+          // Returns the error message as String or null if it succeeded.
+          Function fn = (String name) async {
```
Comment (mkustermann) [inline app/test/backend_test.dart]: The
```dart
Function fn = (String name) async {
```
can be simplified to 
```dart
fn(String name) async {
```

File: app/lib/utils.dart [hunk 4]
```diff
@@ -51,3 +51,56 @@ String canonicalizeVersion(String version) {
   }
   return canonicalVersion.toString();
 }
+
+RegExp _identifierExpr = new RegExp(r'^[a-zA-Z0-9_]*$');
```
Comment (mkustermann) [inline app/lib/utils.dart]: Replace `*` by `+`, we already know it's not the empty string (which is handled before).

**Comment by mkustermann** [comment]
Looks fine with me. Added a few comments. 

Would be **really nice** to share the entire logic of `pubspec.yaml` validation with the client. (or have at least a library we can copy entirely)

**Comment by isoos** [comment]
Done w/ the requested changes. Re: `pubspec.yaml` validation: we could either depend on the `pub` package, or create a new one (under `/dart-lang/`, I think a Googler should own it). Either way, I can do the PRs to share the same codebase.

If you don't mind, I'll open a new PR with more rigorous dart analyzer settings (e.g. finals enforced).

File: app/test/backend_test.dart [hunk 5]
```diff
@@ -731,6 +731,37 @@ main() {
           });
         });
 
+        scopedTest('bad package names are rejected', () async {
+          var tarballStorage = new TarballStorageMock();
```
Comment (mkustermann) [inline app/test/backend_test.dart]: nit: These variables could also be final. Just for future CLs.

**Pull request merged**